lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,486 @@
+            return b;
+        }
+    } else if (operator === "===" || operator === "!==") {
+        if (
+            (isNullOrUndefined(a) && hasConstantNullishness(b)) ||
+            (isStaticBoolean(a) && hasConstantStrictBooleanComparison(b))
+        ) {
+            return b;
+        }
+    }
+    return null;
+}
+
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+/** @type {import('../shared/types').Rule} */
+module.exports = {
+    meta: {
+        type: "problem",
+        docs: {
+            description: "disallow expressions where the operation doesn’t affect the value",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/no-constant-binary-expression"
+        },
+        schema: [],
+        messages: {
+            constantBinaryOperand: "Unexpected constant binary expression. Compares constantly with the {{otherSide}}-hand side of the `{{operator}}`.",
+            constantShortCircuit: "Unexpected constant {{property}} on the left-hand side of a `{{operator}}` expression.",
+            alwaysNew: "Unexpected comparison to newly constructed object. These two values can never be equal."
+        }
+    },
+
+    create(context) {
+        return {
+            LogicalExpression(node) {
+                const { operator, left } = node;
+
+                if ((operator === "&&" || operator === "||") && hasConstantTruthiness(left)) {
+                    context.report({ node: left, messageId: "constantShortCircuit", data: { property: "truthiness", operator } });
+                } else if (operator === "??" && hasConstantNullishness(left)) {
+                    context.report({ node: left, messageId: "constantShortCircuit", data: { property: "nullishness", operator } });
+                }
+            },
+            BinaryExpression(node) {
+
+                const { right, left, operator } = node;
+                const rightConstantOperand = findBinaryExpressionConstantOperand(left, right, operator);
+                const leftConstantOperand = findBinaryExpressionConstantOperand(right, left, operator);
+
+                if (rightConstantOperand) {
+                    context.report({ node: rightConstantOperand, messageId: "constantBinaryOperand", data: { operator, otherSide: "left" } });
+                } else if (leftConstantOperand) {
+                    context.report({ node: leftConstantOperand, messageId: "constantBinaryOperand", data: { operator, otherSide: "right" } });
+                } else if (operator === "===" || operator === "!==") {
+                    if (isAlwaysNew(left)) {
+                        context.report({ node: left, messageId: "alwaysNew" });
+                    } else if (isAlwaysNew(right)) {
+                        context.report({ node: right, messageId: "alwaysNew" });
+                    }
+                }
```

Comment:
Would it be possible to also check `==` and `!=`?

Something like:

```js
} else if (operator === "==" || operator === "!=") {
    if (isAlwaysNew(left) && isAlwaysNew(right)) {
        // report
    }
}
```

Reply:
This would work for this specific case, and would be easy to implement. I think it's probably worth adding.

If we wanted to generalize this approach, we would need to define some helpers which try to statically determine the type (object/string/number/null/undefined) of a given expression. We could then combine the knowledge of the type of each side of the comparison in order to determine if their comparison would always be static. However, this is significantly more involved, so I recommend we don't take that on, or at the very least defer it to a separate project.

Reply:
Yeah, I don’t think it’s worth trying to add type detection here (that would need an RFC and would probably be better off as part of eslint-scope). Let’s just stick with the simple solution for this PR

Reply:
I've updated the code to cover this case.